.. _references-label:

References
==========

**Plant Data Visualization Tool Bundles**

* **ELIXIR bio.tools** (pending)
* **Google OpenDocs** (pending)

**Bundle Tools (cork oak data compatible)**

* **Mercator4** - `Bolger, M., Schwacke, R., & Usadel, B. (2021). MapMan Visualization of RNA-Seq Data Using Mercator4 Functional Annotations. In D. Dobnik, K. Gruden, Ž. Ramšak, & A. Coll (Eds.), Solanum tuberosum: Methods and Protocols (pp. 195–212). Springer US. https://doi.org/10.1007/978-1-0716-1609-3_9 <https://link.springer.com/protocol/10.1007/978-1-0716-1609-3_9>`_
* **MapMan App** - `Schwacke, R., Ponce-Soto, G. Y., Krause, K., Bolger, A. M., Arsova, B., Hallab, A., Gruden, K., Stitt, M., Bolger, M. E., & Usadel, B. (2019). MapMan4: A Refined Protein Classification and Annotation Framework Applicable to Multi-Omics Data Analysis. Molecular Plant, 12(6), 879–892. https://doi.org/10.1016/j.molp.2019.01.003 <https://www.sciencedirect.com/science/article/pii/S1674205219300085?via%3Dihub>`_
* **DiNAR** - `Zagorščak, M., Blejec, A., Ramšak, Ž., Petek, M., Stare, T., & Gruden, K. (2018). DiNAR: Revealing hidden patterns of plant signalling dynamics using Differential Network Analysis in R. Plant Methods, 14(1), 78. https://doi.org/10.1186/s13007-018-0345-0 <https://plantmethods.biomedcentral.com/articles/10.1186/s13007-018-0345-0>`_
* **Ensembl Plants** - `Ensembl Plants Website <https://plants.ensembl.org/index.html>`_ 

**Additional Tools**

* **Seidr** - `Schiffthaler, B., Zalen, E. van, Serrano, A., Street, N., & Delhomme, N. (2021). Seiðr: Efficient Calculation of Robust Ensemble Gene Networks. <https://doi.org/10.1101/250696 <https://seidr.readthedocs.io/en/latest/>`_
* **Cytoscape** - `Shannon, P., Markiel, A., Ozier, O., Baliga, N. S., Wang, J. T., Ramage, D., Amin, N., Schwikowski, B., & Ideker, T. (2003). Cytoscape: A Software Environment for Integrated Models of Biomolecular Interaction Networks. Genome Research, 13(11), 2498–2504. https://doi.org/10.1101/gr.1239303 <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC403769/>`_


**Biological Data**

* **Cork Oak Co-expression Network** - `Barros, P. M., Sapeta, H., Lucas, D., Rodrigues, H., & Oliveira, M. (2022, July 7). Identification of cork oak genetic modules regulating phellem development in response to drought. https://doi.org/10.5281/zenodo.7225240 <https://zenodo.org/record/7225240#.ZDWB5ObMJPY>`_
* **Cork Oak Seasonal Growth** - `Fernández-Piñán, S., Boher, P., Soler, M., Figueras, M., & Serra, O. (2021). Transcriptomic analysis of cork during seasonal growth highlights regulatory and developmental processes from phellogen to phellem formation. Scientific Reports, 11(1), 12053. https://doi.org/10.1038/s41598-021-90938-5 <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8187341/>`_



